### PR TITLE
Protected build linker fixes

### PR DESCRIPTION
**Pull request by jlaitine**
This is a collection of cherry-picks from the "protected build" PR #16931 , in attempt to minimize that one.

This changes linking order of some of the libraries, adds some missing dependencies and removes some circular dependencies.

The linking for protected build is crucial part, since it needs to be done very carefully for kernel and user side; in user side one cannot link libraries which access HW or OS structured directly. Also libraries which are used in both sides cannot be directly linked to HW dependent ones

This is just a bunch of build script changes, it shouldn't change the functionality for any targets.

### CODE REVIEW

**Comment by jlaitine** [comment]
Cleaned up some of the commits, now it should be ok

File: src/lib/drivers/device/CMakeLists.txt [hunk 1]
```diff
@@ -53,8 +53,13 @@ px4_add_library(drivers__device
...
-	target_link_libraries(drivers__device PRIVATE nuttx_arch)
+	if (NOT DEFINED CONFIG_BUILD_FLAT)
+	  target_link_libraries(drivers__device PRIVATE nuttx_karch)
+	else()
+	  target_link_libraries(drivers__device PRIVATE nuttx_arch)
+	endif()
 endif()
 
-target_link_libraries(drivers__device PRIVATE cdev)
+target_link_libraries(drivers__device PRIVATE px4_work_queue)
```
Comment (dagar) [inline src/lib/drivers/device/CMakeLists.txt:65]: This one surprises me.

**Review by dagar — APPROVED** [review]

**Comment by dagar** [comment]
Thanks for bringing this in incrementally @jlaitine.

File: src/lib/drivers/device/CMakeLists.txt (hunk 1 above)
Comment (jlaitine) [reply src/lib/drivers/device/CMakeLists.txt:65]: This looks very weird to me as well. I will check and revert this line if it is wrong - it certainly does look wrong, I don't remember why or where this came from...

Comment (jlaitine) [reply src/lib/drivers/device/CMakeLists.txt:65]: This was indeed some fumble in re-basing commits. I reverted this line separately in #19092 . Other changes in the same file were as intended

**Pull request merged**